pkg/querier/frontend/frontend.go

```diff
@@ -179,6 +179,16 @@ func (f *Frontend) Handler() http.Handler {
 }
 
 func (f *Frontend) handle(w http.ResponseWriter, r *http.Request) {
+	// to parse form we need to be sure
+	// that roundtriper request gets not read reader
+	var buf bytes.Buffer
+	r.Body = ioutil.NopCloser(io.TeeReader(r.Body, &buf))
+	// Ensure the form has been parsed so all the parameters are present
+	err := r.ParseForm()
+	if err != nil {
+		level.Warn(util.WithContext(r.Context(), f.log)).Log("msg", "unable to parse form for request", "err", err)
```

Comment:
Answered in https://github.com/cortexproject/cortex/pull/3276#discussion_r500426485

Reply:
By continuing here, `buf` is in an unknown state... it can be completely empty (eg. for `POST` request with content type other than `application/x-www-form-urlencoded`, and wrong query string), or filled with broken / unfinished `x-www-form-urlencoded` data). In either case, it's different than it was before parsing the form.

Reply:
Added return statements in case of errors, also added check for `_, err = io.Copy(w, resp.Body)`